typing/typecore.ml

```diff
@@ -990,13 +968,134 @@ let rec has_literal_pattern p = match p.ppat_desc with
+      generated by the pattern-matching analyser. *)
+
+
+(** In [Counter_example] mode the responsibility of [type_pat] is to
+   check that a counter-example to exhaustivity is valid: it has to be
+   type-correct in a slightly different sense than for user-code, see
+   [splitting_mode] below, and also capture at least one concrete
+   value.
+
+   [counter_example_checking_info] contains various parameters for
+   this mode of operation. *)
+and counter_example_checking_info = {
+    explode: explosion_fuel;
+    splitting_mode: splitting_mode;
+    synthetic_names: counter_example_synthetic_names;
+  }
+
+
+(** For a pattern to be a valid counter-example, it must be
+   inhabited -- match at least one value. A pattern may fail to match
+   a value if some of its wildcard are at a type that is empty
+   (typically due to incompatible GADT equations, although we also
+   support empty sum types).
+
+   The type-checker tries to figure out if a wildcard is in fact empty
+   by "exploding" it, doing a case-analysis on possible
+   constructors. It only tries up to a fixed depth bound, to avoid
+   non-termination (this is a generic proof-search problem) or
+   compilation-time slowdowns.
+
+   [explosion_fuel] is this depth limit on wildcard explosion. *)
+and explosion_fuel = int
+
+
+(** A counter-example pattern may not be well-typed because the
+   generator does not know about GADTs. In particular, it may generate
+   (even in depth) disjunctions of the form
+
+         (p1 | p2 | ... | pn),
+
+   where some of the pi may in fact be ill-typed. If at least one of
+   them is well-typed, we do have a valid counter-example. So we want
+   the type-checker to succeed if at least one of the alternatives is
+   well-typed.
+
+   Note that the choice of value in such an or-pattern can influence
+   the validity of other parts of the counter-example: if the
+   alternatives are GADT constructors, each one may add equations that
+   restrict other part of the counter-example in its own way. As
+   a consequence, evaluating each choice may require arbitrary
+   search/filtering in other parts of the counter-example (in
+   implementation terms, the continuation [k] is called on each
+   alternative).
+
+   Searching after each alternative in this way may be very costly. It
+   is only useful when GADTs are involved inside the or-pattern; if no
+   GADT constructors occur (possibly deep) inside an alternative, it
+   does no restrict the rest of the search space, so we can go with
+   this alternative and never try the others. But we don't know
+   whether a GADT constructor will occur (we haven't checked the
+   pattern yet!).
+
+   A possible implementation is to optimistically assume that
+   splitting will not be necessary, and split only if we hit a GADT
+   under an or-pattern. This strategy is faster when GADTs are rare.
+
+   The [splitting_mode] setting determines the splitting policy.  *)
+and splitting_mode =
+  | Pessimistic
```

Comment:
I'm happy to get rid of the `Split_or / Pessimistic` mode in the future and use `Normal / Optimistic` in all situations. But could we maybe try to first have a PR that documents the code without changing the behavior (as the present PR is designed to be), and then change the behavior (if we want to make changes) in a separate PR?

1. Having behavior-changes as separate PRs would make it easier to spot and analyze regression if some were to come up unexpectedly.
2. Modifying the clarified/documented code is nicer to do (and review) than modifying the obfuscated version.
3. We have discussed some other potential changes (I pointed out that there seems to be no real need to `raise Need_backtrack` in the `Ppat_any` case as is currently done) that are waiting for further PRs anyway.

If we agree on this, then what should we change to this documentation? The current writing suggests that Pessimistic is used on refutation clauses because it uses GADTs. (This is a justification that Florian and myself reverse-engineered -- made up.) Would you prefer me to write that we don't know why Pessimistic is used, because it sounds like a worse choice than Optimistic in all situations? (I now agree with that.)

Reply:
I think the real reason is in the second part of my answer. So it should probably be `Refine_or` / `Backtrack_or`, or something equivalent.